tokio-io/src/split.rs

```diff
@@ -0,0 +1,175 @@
+//! Split a single value implementing `AsyncRead + AsyncWrite` into separate
+//! `AsyncRead` and `AsyncWrite` handles.
+//!
+//! To restore this read/write object from its `split::ReadHalf` and
+//! `split::WriteHalf` use `unsplit`.
+
+use crate::{AsyncRead, AsyncWrite};
+
+use bytes::{Buf, BufMut};
+use futures_core::ready;
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::sync::atomic::AtomicU8;
+use std::sync::atomic::Ordering::{Acquire, Release};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+/// The readable half of a value returned from `split`.
+pub struct ReadHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+/// The writable half of a value returned from `split`.
+pub struct WriteHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+struct Inner<T> {
+    state: AtomicU8,
+    stream: UnsafeCell<T>,
+}
+
+struct Guard<'a, T> {
+    inner: &'a Inner<T>,
+}
+
+const INIT: u8 = 0;
+const LOCKED: u8 = 1;
+
+/// Split a single value implementing `AsyncRead + AsyncWrite` into separate
+/// `AsyncRead` and `AsyncWrite` handles.
+///
+/// To restore this read/write object from its `split::ReadHalf` and
+/// `split::WriteHalf` use `unsplit`.
+pub fn split<T>(stream: T) -> (ReadHalf<T>, WriteHalf<T>) {
+    let inner = Arc::new(Inner {
+        state: AtomicU8::new(INIT),
+        stream: UnsafeCell::new(stream),
+    });
+
+    let rd = ReadHalf {
+        inner: inner.clone(),
+    };
+
+    let wr = WriteHalf { inner };
+
+    (rd, wr)
+}
+
+impl<T> ReadHalf<T> {
+    /// Reunite with a previously split `WriteHalf`.
+    ///
+    /// # Panics
+    ///
+    /// If this `ReadHalf` and the given `WriteHalf` do not originate from the
+    /// same `split` operation this method will panic.
+    pub fn unsplit(self, wr: WriteHalf<T>) -> T {
+        if Arc::ptr_eq(&self.inner, &wr.inner) {
+            drop(wr);
+
+            let inner = Arc::try_unwrap(self.inner)
+                .ok()
+                .expect("Arc::try_unwrap failed");
+
+            inner.stream.into_inner()
+        } else {
+            panic!("Unrelated `split::Write` passed to `split::Read::unsplit`.")
```

Comment:
I feel we should just return a result here? but this is fine for now.

Reply:
My philosophy: `Result` for errors that can naturally occur at runtime, `panic` for bugs that can be fixed by the caller.

Reply:
I could see someone maybe making a runtime mistake that then panics? But now that I think about that it is pretty fatal. 🤷‍♂